### DISCUSSION TITLE
Will TTL play a role in cleaning up old values if replace before TTL?

### DISCUSSION THREAD

**Discussion by ShanmugamC** [ben-manes/caffeine#570] [Q&A] [answered]
Hi Team,
              If we set higher TTL for a key and end up replacing value for it before TTL is over multiple times, what happens to old values as i see some memory leak for this scenario?

E.g:
   Key 1:  TTL : 12 hours but i replace the value 100 time within 12 hours,  what happens in this case? Will it hold reference for old values still?

**Comment by ben-manes** [comment, marked as answer]
This depends on your expiration setting.
- `expireAfterWrite` will reset the entry's lifetime on every update, causing it to have a 12 hour duration.
- `expireAfterAccess` will reset the entry's lifetime on every read or write.
- `expireAfter(expiry)` will reset to the return value of `Expiry.expireAfterUpdate`.

If you want an entry to expire only after creation, regardless of subsequent reads or writes, then you should use `expireAfter(expiry)`.
```java
Caffeine.newBuilder()
    .expireAfter(new Expiry<Key, Graph>() {
      public long expireAfterCreate(Key key, Graph graph, long currentTime) {
        return TimeUnit.HOURS.toNanos(12);
      }
      public long expireAfterUpdate(Key key, Graph graph, 
          long currentTime, long currentDuration) {
        return currentDuration;
      }
      public long expireAfterRead(Key key, Graph graph,
          long currentTime, long currentDuration) {
        return currentDuration;
      }
    }).build();
```

**Reply by ShanmugamC** [reply]
sure , Thanks for your detail reply. 

I noticed memory leak during this process as i mentioned in the question earlier, so my question is if i override the same key , will it still hold reference with older value and cause memory leak?

**Reply by ben-manes** [reply]
Oh, sorry. I misread that then.

The old value is removed. If you have a listener configured then it will be notified. The cache is merely a fancy `ConcurrentHashMap` so it should follow those expectations.

**Reply by ShanmugamC** [reply]
Thanks for your reply @ben-manes  !

**Reply by ShanmugamC** [reply]
I am Initiating cache as below:

// Class CacheRegion holds below

           AsyncCache<K, V> asyncCache = Caffeine.newBuilder()
                    .maximumSize(5000)
                    .expireAfterWrite(Duration.ofHours(12))
                    .recordStats()
                    .buildAsync();

//And i am overriding same cache multiple times before cache expiring using below code with spring webflux

      CompletableFuture<CacheObject<T>> cf = CompletableFuture.completedFuture(cacheObject);
      asyncCache.put(key, cf);

// I was able to reproduce CacheRegion holding lot of memory for every-time i override the same key. With 12 hours TTL slowly memory builds up and causing OOM , but to confirm this i reduced TTL to 30 mins and after 30 mins JVM memory usage is stable and i don't see the issue.

Just wanted to understand if i am doing something wrong here and why old keys/values still stay after overriding with new key & value.

**Reply by ben-manes** [reply]
That looks correct. The cache internally is a `Map<K, CompletableFuture<V>>` and it won't hold onto the old value.

fyi, you can use `asyncCache.synchronous.put(key, cacheObject)` which wraps the value for you.
